module/Finna/src/Finna/View/Helper/Root/CustomElement.php

```diff
@@ -0,0 +1,99 @@
+<?php
+/**
+ * Custom element view helper
+ *
+ * PHP version 7
+ *
+ * Copyright (C) The National Library of Finland 2021.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 2,
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * @category VuFind
+ * @package  View_Helpers
+ * @author   Aleksi Peebles <aleksi.peebles@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+namespace Finna\View\Helper\Root;
+
+use Finna\View\CustomElement\CustomElementRendererInterface;
+use Finna\View\CustomElement\PluginManager as CustomElementManager;
+use Laminas\ServiceManager\Exception\ServiceNotFoundException;
+use Laminas\View\Helper\AbstractHelper;
+
+/**
+ * Custom element view helper
+ *
+ * @category VuFind
+ * @package  View_Helpers
+ * @author   Aleksi Peebles <aleksi.peebles@helsinki.fi>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+class CustomElement extends AbstractHelper implements CustomElementRendererInterface
+{
+    /**
+     * Custom element plugin manager
+     *
+     * @var CustomElementManager
+     */
+    protected $pluginManager;
+
+    /**
+     * Constructor
+     *
+     * @param CustomElementManager $pm Custom element plugin manager
+     */
+    public function __construct(CustomElementManager $pm)
+    {
+        $this->pluginManager = $pm;
+    }
+
+    /**
+     * Server-side render a custom element.
+     *
+     * @param string $name    The name of the custom element to render
+     * @param ?array $options Options to use when creating the element instance
+     *
+     * @return ?string Rendered element, or null if element does not exist
+     */
+    public function __invoke(string $name, ?array $options = null): ?string
+    {
+        return $this->render($name, $options);
+    }
+
+    /**
+     * Server-side render a custom element.
+     *
+     * @param string $name    The name of the custom element to render
+     * @param ?array $options Options to use when creating the element instance
+     *
+     * @return ?string Rendered element, or null if element does not exist
+     */
+    public function render(string $name, ?array $options = null): ?string
+    {
+        try {
+            $options = $options ?? [];
+            $options['element'] = $name;
```

Comment:
Onko ok, että elementin tagi määräytyy pyydetyn nimen perusteella? Nimihän on periaatteessa vain alias, ja hieman mietityttää se, että alias määrittää tagin. Ainakin tagista tulisi vähän kumma, jos pyytäisi elementtiä luokan nimellä. Eihän siinä järkeä ole, mutta voisi ehkä tarkistaa, että `$name` on 'finna-jotain'?

Reply:
Ongelma on siis se, että alias on vain alias, ja luokalla voi periaatteessa olla aliaksia vaikka kuinka monta. Silloin tavallaan tagin määrittelyn omistajuus on epäselvä. Aliakset ovat lisäksi case-insensitiivisiä, jolloin voi pyytää vaikka Finna-pAnEL:n.

Reply:
HTML-taginimet ovat myös case-insensitiivisiä, eli on itse asiassa hyvä juttu että aliakset toimivat noin. Sekään ei välttämättä ole huono asia, että saman toiminnallisuuden saisi useilla tageilla. Määrittelyn omistajuus on sitten haastavampi, jos ei ole mahdollista kumota jo asetettua aliasta? Eli pitäisikö siis tehdä oma config-tiedosto? Erillinen vai osaksi olemassa olevaa?

Reply:
Näemmä mahdollista, mutta toki vähän turhan hankalaa: https://docs.laminas.dev/tutorials/advanced-config/#manipulating-merged-configuration

Reply:
Jos elementti ei millään voi itse määrätä tagiaan, niin en sitä ryhtyisi mihinkään erilliseen konfiguraatioonkaan laittamaan vaan tyytyisin nykytilanteeseen. Se on vaan epäselvä, siksi en tykkää siitä, mutta parempi silti niin kuin vielä jossain erikseen.

Reply:
Elementti voi toki määrätä taginsa, kuten ensimmäisessä PR-versiossa oli, mutta ongelma on kai siinä, että sen pitäisi pystyä määräämään, millä nimellä sitä service containerilta pyydetään? Ja näin ei kai pysty aukottomasti tekemään, vaikka ensimmäisen PR-version metodikutsulla haetulla aliaksella tätä vähän tavoiteltiinkin?

Jos tavoitteena on että parseri saa jollain tapaa pyydettyä luokan, joka osaa prosessoida parserin löytämän tagin, niin  service containerin ja pluginien käyttö kai on luontevin tapa, muutenhan pitäisi itse toisintaa käytännössä sama toiminnallisuus (esim. config-osion avulla)? Vai olisiko tähän joku yksinkertaisempi tapa jolla saman asian voisi tehdä?